## Introduction to Data structures <a name="datastructures"></a>

While reading, pay attention to the types of brackets.

**Lists** and arrays are the same thing in Python. https://docs.python.org/3/tutorial/introduction.html#lists and https://docs.python.org/3/tutorial/datastructures.html#more-on-lists

In [13]:
vaccination_queue = ['Summer', 'Jerry', 'Beth', 'Rick', 'Morty']
type(vaccination_queue)

list

In [2]:
import numpy as np
val = [1, 2, 3, 2022]
print(type(val))
num_val = np.array(val)
print(val)
print(num_val, type(num_val))

<class 'list'>
[1, 2, 3, 2022]
[   1    2    3 2022] <class 'numpy.ndarray'>


In [3]:
# Python keeps the elements of list in the same order
vaccination_queue

['Summer', 'Jerry', 'Beth', 'Rick', 'Morty']

**Tuple**

https://docs.python.org/3/tutorial/datastructures.html#tuples-and-sequences

In [1]:
countries_with_a = ('Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola', 'Antigua and Barbuda')
type(countries_with_a)

tuple

In [2]:
countries_with_a

('Afghanistan',
 'Albania',
 'Algeria',
 'Andorra',
 'Angola',
 'Antigua and Barbuda')

Let's replace one of the countries:

In [3]:
countries_with_a[0] = 'Another country'


TypeError: 'tuple' object does not support item assignment

The **TypeError** was raised because, unlike lists, tuples are immutable. This is why we cannot change the assigned items.

### For loop

In [4]:
for i in countries_with_a:
    print(i)

Afghanistan
Albania
Algeria
Andorra
Angola
Antigua and Barbuda


In [5]:
for i in range(len(countries_with_a)):
    print(i)

0
1
2
3
4
5


In [6]:
for i in range(1, len(countries_with_a) - 1, 2):
    print(i)
#where is number 5?

1
3


In [ ]:
for idx, country in enumerate(countries_with_a):
    print(idx, country)

### While loop

In [ ]:
counter = 1
while counter < 6:
    print(counter)
    counter += 1

In [7]:
countries_with_a = list(countries_with_a)
counter = len(countries_with_a)

while counter > 0:
    print(countries_with_a[-1])
    countries_with_a.pop()
    counter -= 1
else:
    print("There are no countries left on the list")

Antigua and Barbuda
Angola
Andorra
Algeria
Albania
Afghanistan
There are no countries left on the list


## Indexing and slicing <a name="indexing"></a>

In [ ]:
vaccination_queue

In [ ]:
# list length
len(vaccination_queue)

Each item in the list has its own index. 0 - 'Summer', 1 - 'Jerry', etc. You can return an item from a list by its index using square brackets [ ].

<img src="https://cdn.programiz.com/sites/tutorial2program/files/python-list-index.png" width="500" height="500" align="left">

Mind that indexing in Python starts at 0 instead of 1. 

Task! Check who is the first in a line,

In [ ]:
# YOUR CODE

Who is the last one?

In [ ]:
# YOUR CODE


or

In [ ]:
# the index of the last element is (length - 1) because indices start at 0 instead of 1

vaccination_queue[len(vaccination_queue) - 1]

Explanation: **[start index : end index : step]** [more details](https://stackoverflow.com/questions/509211/understanding-slice-notation)

['Summer', 'Jerry', 'Beth', 'Rick', 'Morty']

    0      1    2     3    4

In [ ]:
# all elements starting from the second element
vaccination_queue[1:]

In [ ]:
# all elements before the last element
vaccination_queue[:-1]

In [ ]:
# all elements with one step skip (keep elements at even indices only)
vaccination_queue[::2]

In [ ]:
# all elements in reverse order
vaccination_queue[::-1]

__Slicing__
`slice(start, end, step)`

In [17]:
print(vaccination_queue)
vaccination_queue[slice(1,3,1)]

['Summer', 'Jerry', 'Beth', 'Rick', 'Morty']


['Jerry', 'Beth']

### Lists methods

In [ ]:
# alternative way to reverse list in place
vaccination_queue.reverse()
vaccination_queue

In [ ]:
# nested list indexing
vaccination_queue.append(['Bad Snuffles', 'Good Snuffles', 'Alternative Snuffles'])
vaccination_queue

In [ ]:
vaccination_queue.extend(['Bad Snuffles', 'Good Snuffles', 'Alternative Snuffles'])
vaccination_queue

__TEST__ `.append()` and `.extend()` methods yourself

In [ ]:
# YOUR CODE

**Task!** How to access to the 'Alternative Snuffles'?

In [ ]:
# YOUR CODE

In [18]:
# Grades for Exam tasks from 1 to 5
your_potential_grades = [5, 4.5, 5, 4.2, 5]

# Look up your second mark (Task-2)
your_potential_grades[1] 

4.5

In [ ]:
arr = [your_potential_grades[i] for i in range(len(your_potential_grades))]
print(arr)

In [ ]:
weight = [0.4, 0.1*0.6, 0.1*0.6, 0.15*0.6, 0.25*0.6]
grade_before_rounding = 0
for i in range(len(weight)):
    grade_before_rounding += weight[i] * arr[i]
print(grade_before_rounding)

Let's sort the grade **in ascending order** in place (it means, that this list will be stored as sorted). 

In [ ]:
# sorting 

your_potential_grades.sort()
your_potential_grades

In [ ]:
# click inside the brackets of sort and push Shift + Tab
your_potential_grades.sort()

This is how you call the documentation for each function in Python. So, you can see, there are two arguments: key and reverse. Let's use reverse and sort our list **in descending order**.

In [19]:
your_potential_grades.sort(reverse=True)
your_potential_grades

[5, 5, 5, 4.5, 4.2]

In [20]:
# alternative with creteing a copy of the list

your_potential_grades = [5, 4.5, 5, 4.2, 5]

sorted(your_potential_grades)

[4.2, 4.5, 5, 5, 5]

In [21]:
your_potential_grades

[5, 4.5, 5, 4.2, 5]

Why didn't Python keep your_potential_grades sorted and return the original order of the elements? This is because sorted () creates a copy of your list and then sorts it.

In [ ]:
sorted_grades = sorted(your_potential_grades)
sorted_grades

Your can change the grade you don't like, because lists are mutable. 

In [ ]:
# replace
print(your_potential_grades[1])
your_potential_grades[1] = 5
your_potential_grades[1]

Imagine, you like this course so much, that you even ask additional special tasks and get 2.5 additional points. 

In [ ]:
# add element 
your_potential_grades.append(2.5)

your_potential_grades

> __Unfortunately, the rules does not allow to give you additional points__

In [ ]:
# delete element
your_potential_grades.remove(2.5)

your_potential_grades

In [ ]:
# pop element
your_potential_grades.pop()

In [ ]:
your_potential_grades

pop() returns the last item in the list and removes it, while the remove() method removes the item you specified in brackets (with whatever index you want) and does not return that value on output.

In [ ]:
# pop element
your_potential_grades.pop(0)
your_potential_grades

In [ ]:
# you can save the result of pop() into a variable
last_grade_in_list = your_potential_grades.pop()
last_grade_in_list

In [22]:
# insert an element at a specific position
# 0 - index, 9 - value
your_potential_grades.insert(0, 9)
your_potential_grades

[9, 5, 4.5, 5, 4.2, 5]

In [23]:
your_potential_grades.insert(1, 10)
your_potential_grades

[9, 10, 5, 4.5, 5, 4.2, 5]

In [24]:
# let's count how many times the value 10 is in the list
your_potential_grades.count(10)

1

### List comprehensions

In [25]:
some_list = [0, 1, 2, 3, 4, 5, 6]

new_list = []
for i in some_list:
    new_list.append(i ** 2) # square 

new_list1 = [some_list[i] ** 2 for i in range(len(some_list))]
new_list2 = [i ** 2 for i in some_list]
print(new_list, new_list1, new_list2, sep = '\n')

[0, 1, 4, 9, 16, 25, 36]
[0, 1, 4, 9, 16, 25, 36]
[0, 1, 4, 9, 16, 25, 36]


Instead of loop with **for**, we can code this operation in one line using list comprehensions:

In [ ]:
another_new_list = [i ** 2 for i in some_list]
another_new_list

The square brackets around the expression should indicate that we are creating a list. The expression inside the brackets should be read literally: a list of squared elements for (**for**) elements `i` from (**in**) `some_list`.

In [ ]:
# filtering using list comprehensions
[x for x in another_new_list if x % 2 == 0]

% operator returns the remainder of the division. So, ```if x % 2 == 0``` means that if the number x is divisible by 2 and has no remainder (remainder = 0 is True), then we store it in the list, otherwise, delete it. As a result, we get even numbers.

### Tuples

https://www.w3schools.com/python/python_tuples.asp

In [ ]:
info = ('Sherlock Holmes', 'Baker Street 221b', 'Deduction', 'Elementary', 'Dr. Watson')
type(info)

In [ ]:
# access item
info[3]

In [ ]:
print(info)

In [ ]:
info[-1]

In [ ]:
info[-6]

In [ ]:
# slicing
info[2:5]

In [ ]:
info[2:100000000] # the upper bound of index in slicing is unlimited(the same for lists and strings)

In [ ]:
# replacement doesn't work because tuples are immutable
info[2] = 'Induction'
info

However, you can convert your tuple into list, make change and convert it back. 

In [ ]:
info = list(info)
info, type(info)

In [ ]:
info[2] = 'Induction'
info = tuple(info)
info

> **Once a tuple is created, you cannot add and remove items in it**. Tuples are unchangeable, but you can use the trick with list and convert it back.

**DIctionaries**

A dictionary is an unordered structure, it consists of keys and values. We cannot refer to the object by index, but we can refer to the key.

In [26]:
president = {
  "name": "Uknown",
  "age": 81,
  "political party": 'Democratic',
  "children" : ['Hunter', 'Beau', 'Ashley', 'Naomi Christina']
}

In [27]:
'Roman said "something"'

'Roman said "something"'

We cannot refer to the object by index, but we can refer to the key:

In [28]:
president[0]

KeyError: 0

In [29]:
president['age']

81

To find out what keys or values are in the dictionary, you can use the corresponding methods.

In [30]:
president.keys()

dict_keys(['name', 'age', 'political party', 'children'])

In [31]:
print(president.keys()[0])

TypeError: 'dict_keys' object is not subscriptable

In [32]:
print(type(president.keys()))
x = list(president.keys())
print(x[0])

<class 'dict_keys'>
name


In [33]:
president.values()

dict_values(['Uknown', 81, 'Democratic', ['Hunter', 'Beau', 'Ashley', 'Naomi Christina']])

In [34]:
president.items()

dict_items([('name', 'Uknown'), ('age', 81), ('political party', 'Democratic'), ('children', ['Hunter', 'Beau', 'Ashley', 'Naomi Christina'])])

In [35]:
for key, value in president.items():
    print('key:', key, '- value:', value)

key: name - value: Uknown
key: age - value: 81
key: political party - value: Democratic
key: children - value: ['Hunter', 'Beau', 'Ashley', 'Naomi Christina']


In [36]:
for key in president:
    print(key)

name
age
political party
children


Adding key-value pairs to a dictionary is very simple, like replacing a value by index in lists. The deletion is done using the ```del``` function. We can also check the presence of the key in the dictionary.

In [37]:
president['place of birth'] = 'Scranton, Pennsylvania, U.S.' # new key and value
president['education'] = ['University of Delaware (BA)','Syracuse University (JD)']
president['name'] = 'Joe Biden'  # overwrite the value in existing key
print(president)

{'name': 'Joe Biden', 'age': 81, 'political party': 'Democratic', 'children': ['Hunter', 'Beau', 'Ashley', 'Naomi Christina'], 'place of birth': 'Scranton, Pennsylvania, U.S.', 'education': ['University of Delaware (BA)', 'Syracuse University (JD)']}


In [38]:
del president['children']
print(president)

{'name': 'Joe Biden', 'age': 81, 'political party': 'Democratic', 'place of birth': 'Scranton, Pennsylvania, U.S.', 'education': ['University of Delaware (BA)', 'Syracuse University (JD)']}


In [39]:
'place of birth' in president # is 'place of birth' among keys?

True

**Sets**

<img src="https://www.learnbyexample.org/wp-content/uploads/python/Python-Set-Operatioons.png" height=300 width=300>

In [ ]:
european_union = {"Austria", "Belgium", "Bulgaria", "Croatia", "Cyprus", "Czech Republic", "Denmark",
    "Estonia", "Finland", "France", "Germany", "Greece", "Hungary", "Ireland", "Italy", "Latvia", 
    "Lithuania", "Luxembourg", "Malta", "Netherlands", "Poland", "Portugal", "Romania", 
    "Slovakia", "Slovenia", "Spain", "Sweden"}

eurozone = {"Austria", "Belgium", "Croatia", "Cyprus", "Portugal", "Spain", 
            "Slovakia", "Slovenia", "Ireland", "Italy", "France", "Germany", "Greece",
            "Estonia", "Finland", "Latvia", "Lithuania", "Luxembourg", "Malta", "Netherlands"}

In [ ]:
print(european_union | eurozone)  # union of sets -- we join all unique names from two sets together
print(european_union & eurozone)  # intersection -- keep the elements that are in both sets
print(european_union - eurozone)  # difference -- elements in the first set, but not in the second
print(european_union ^ eurozone)  # symmetric difference -- elements from A | B, but not from A & B

Instead of signs as operators you can use methods:

In [ ]:
print(european_union.union(eurozone))
print(european_union.intersection(eurozone))
print(european_union.difference(eurozone))
print(european_union.symmetric_difference(eurozone))

Summing up the properties. **Ordered** means that Python stores elements in the same order you specified when you created it. **Mutable** means that we can change the elements, for example, to replace one of the elements with a new value. **Constructor** demonstrates how to create, represent and store each data type. Note the difference in brackets.

<img src="https://miro.medium.com/max/1400/1*Det-kkoSw9T4IZ4XrypVNQ.png" style="width:700px;height:300px"/>

## Exercises

**Task 1.** <br>
1.1 Create a dictionary with the countries as the keys and information on the population (in millions) as the values. The order is preserved (the first value in ```population_mln``` is for the first country in the list ```countries```, the second value - for the second country, etc.). <br>
1.2 Add the new values in this dictionary: Germany - 83, France - 67, China - 140.81. <br>
1.3 Oops! We made a mistake for China. It's 1408.1 million. Replace the existing value with the correct one.

In [ ]:
countries = ['Italy', 'Poland', 'Austria', 'Australia', 'Cyprus', 
             'USA', 'Burkina Faso', 'Spain', 'Ukraine']

population_mln = [60.3, 37.5, 8.6, 25.4, 0.86, 328, 20.3, 47, 43.8]


# YOUR CODE HERE

**Task 2.** Write the code to return the second maximum value in the list. This means that it is less than the maximum value, but greater than the others.

In [ ]:
array = [59, 7, 9, 87, 1, 0, 2, 20, 62, 120, 121, 98]

# YOUR CODE HERE

#sort and print the second element
#time costly, as it uses O(n*log(n))

#think about alternatives

## Strings

The key operations: concatenation, split, join and extending string with its duplicates by *N* times. 

In [41]:
greeting = 'welcome to the course on Software Engineering!'

whom = 'Dear students,\n\n'

# concatenation
print(whom + greeting)

Dear students,

welcome to the course on Software Engineering!


In [40]:
# repeat by mulyiplicator
print('You will need to work', 'very ' * 5, 'hard...')

You will need to work very very very very very  hard...


In [42]:
# verify whether the string starts with
print(greeting.startswith('Chilling'))

False


In [48]:
# string ends with
print('"greeting":', greeting)
print(greeting.endswith('sis!'))
print('Recheck it'.endswith('t'))

"greeting": welcome to the course on Software Engineering!
False
True


In [ ]:
# split
text_line = 'You need_to split this_sentence into four parts_by underscore sign.'
print(text_line.split('_'))

In [ ]:
type(text_line.split('_'))

In [ ]:
# join
text_line = 'You need to join this sentence by asterisks instead of white spaces.'
'*'.join(text_line.split(' '))

The find( ) method returns the index of the first character of the first occurrence of the substring, if contained in the string. If there is no such substring, it returns -1.

In [49]:
# find

prologue = '''The morning had dawned clear and cold, with a crispness that hinted at
the end of summer. They set forth at daybreak to see a man beheaded, twenty in all, and
Bran rode among them, nervous with excitement. This was the first time he had been deemed
old enough to go with his lord father and his brothers to see the king’s justice done. It was
the ninth year of summer, and the seventh of Bran’s life'''

prologue.find('Bran')

159

The name 'Bran' was found and 159 is the index of the first letter of this word in a string 'prologue'. 

In [50]:
print(prologue[prologue.find('Bran'):]) # prints the whole string starting from 'Bran'

Bran rode among them, nervous with excitement. This was the first time he had been deemed
old enough to go with his lord father and his brothers to see the king’s justice done. It was
the ninth year of summer, and the seventh of Bran’s life


In [51]:
print(prologue[prologue.find('Bran'):prologue.find(' rode')]) #slicing

Bran


Let's make a simple version of sentiment analysis from scratch. If the review contains the word 'bad', then we will write that this is a negative review.

In [ ]:
feedback = 'This place was bad enough.'

print(feedback)
if feedback.find('bad') == -1: # if 'bad' not found
    print('Feedback is not negative')
else:
    print('Feedback is negative')

In [ ]:
# replace
statement = '''I take you, Monica, to be my wife.\
I promise to be true to you in good times and in bad, in sickness and in health.\
I will love you and honour you all the days of my life.'''
#This is Traditional Wedding Vows

statement.replace('Monica', 'Rachel')

__Exercise__

Write a code that find all entries of substring <code>t</code> in the string <code>s</code> including the intersection. The output is the list of indexes, where each of them corresponds to the first character of <code>t</code>. The code returns <code>None</code> if <code>s</code> is absent

**Task 1. Hangman game** 

You need to write a program for the Hangman word game.

One player comes up with a word to guess and uses input() to save it. Only lowercase letters are allowed.

The other player must guess the word by suggesting the letters in it. The user provides each new letter using the input() function. Only 10 attemps are allowed. If the word contains a letter, print "Correct!". If the letter is incorrect, print "There is no such letter!". *Hint: use ```word_set``` to check if the guessed letter is in the word or not.*

If all the letters from the word are named before the attempts have ended, you should print “Bingo! The word was X ", where X is the word to guess and stop the game (you can use ```break``` on a separate line with the same indent). *Hint: you can check if a set with named letters equals the set with letters of the initial word.*

## Discussion

__exit from the enternal cycle__: command `break`

In [ ]:
# test break here
i = 0
while True:
    i += 1
    if (i == 5):
        print(i)
        break

In [ ]:
for i in range(100):
    if i == 5:
        print(i)
        break

__Convertion of the word to sets and comparison of sets__

In [ ]:
# compare two sets here

In [52]:
word = 'University'
another_word = 'I love you Monica'
print(set(word), set(another_word), sep = '\n')

{'v', 'n', 'i', 's', 't', 'r', 'e', 'U', 'y'}
{'v', 'n', 'I', 'a', ' ', 'M', 'u', 'o', 'c', 'l', 'e', 'i', 'y'}


In [53]:
word_set = set(word)
print(type(word_set))

<class 'set'>


In [54]:
letter = 'i'
if letter in word_set:
    print('In')
else:
    print('Out')

In


In [ ]:
another_letter = 'z'
if another_letter in word_set:
    print('In')
else:
    print('Out')

## Design of the code

1. Get input word with `input()` and covert it to `set`
2. Define the set of the correct letters as the emptyset
3. Main loop:
    * read a new letter with `input()`
    * if it is correct:
        * place it to the set of the correct letters
        * print the set of the correctly guessed letters
        * in case the sets of the correct letters and the secret word coincides, break the loop with `break` and give the success message
    * else:
        * print the error message
        * print the set of the correctly guessed letters
4. After the loop, print a message if the secret word was not guessed

<div style = "color:red; font-size:36px;"> End of the task with the game </div>
<br><br>